R/listing_export.R

```diff
@@ -0,0 +1,158 @@
+#' main_title(lsting) <- "this is some title"
+#' main_footer(lsting) <- "this is some footer"
+#' cat(export_as_txt(lsting, file = NULL, paginate = TRUE))
+#'
+export_as_txt <- function(lst, file = NULL,
+                          page_type = NULL,
+                          landscape = FALSE,
+                          pg_width = page_dim(page_type)[if (landscape) 2 else 1],
+                          pg_height = page_dim(page_type)[if (landscape) 1 else 2],
+                          font_family = "Courier",
+                          font_size = 8, # grid parameters
+                          paginate = .need_pag(page_type, pg_width, pg_height, lpp, cpp),
+                          cpp = NULL,
+                          lpp = NULL,
+                          ..., page_break = "\\s\\n",
+                          hsep = default_hsep(),
+                          indent_size = 2,
+                          tf_wrap = paginate,
+                          max_width = cpp,
+                          colwidths = propose_column_widths(matrix_form(lst, TRUE))) {
+  if (!is.null(colwidths) && length(colwidths) != ncol(lst)) {
+    stop(
+      "non-null colwidths argument must have length ncol(lst) [",
+      ncol(lst), "], got length ", length(colwidths)
+    )
+  }
+  if (paginate) {
+    gp_plot <- gpar(fontsize = font_size, fontfamily = font_family)
+
+    pdf(file = tempfile(), width = pg_width, height = pg_height)
+    on.exit(dev.off())
+    grid.newpage()
+    pushViewport(plotViewport(margins = c(0, 0, 0, 0), gp = gp_plot))
+
+    cur_gpar <- get.gpar()
+    if (is.null(page_type) && is.null(pg_width) && is.null(pg_height) &&
+      (is.null(cpp) || is.null(lpp))) {
+      page_type <- "letter"
+      pg_width <- page_dim(page_type)[if (landscape) 2 else 1]
+      pg_height <- page_dim(page_type)[if (landscape) 1 else 2]
+    }
+
+    if (is.null(lpp)) {
+      lpp <- floor(convertHeight(unit(1, "npc"), "lines", valueOnly = TRUE) /
+        (cur_gpar$cex * cur_gpar$lineheight))
+    }
+    if (is.null(cpp)) {
+      cpp <- floor(convertWidth(unit(1, "npc"), "inches", valueOnly = TRUE) *
+        font_lcpi(font_family, font_size, cur_gpar$lineheight)$cpi)
+    }
+    if (tf_wrap && is.null(max_width)) {
+      max_width <- cpp
+    }
+
+    tbls <- paginate_listing(lst,
+      lpp = lpp, cpp = cpp,
+      min_siblings = 2,
+      nosplitin = character(),
+      colwidths = propose_column_widths(lst),
+      verbose = FALSE
+    )
+  } else {
+    tbls <- list(lst)
+  }
+
+  res <- paste(
+    mapply(
+      function(tb, ...) {
+        ## 1 and +1 are because cwidths includes rowlabel 'column'
+        # cinds <- c(1, .figure_out_colinds(tb, lst) + 1L)
+        toString(tb, ...)
+      },
+      MoreArgs = list(hsep = hsep),
+      SIMPLIFY = FALSE,
+      tb = tbls
+    ),
+    collapse = page_break
+  )
+
+  if (!is.null(file)) {
+    cat(res, file = file)
+  } else {
+    res
+  }
+}
+
+
+.do_inset <- function(x, inset) {
```

Comment:
This function already exists in formatters, it doesn't belong here. Also, all you should need to do is set the inset on the matrix form, during construction, there's no reason to be doing this here (it does not happen anywhere in rtables, for example)